### DISCUSSION TITLE
Certificates with MQTT and Mosquitto

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#20243] [Q&A] [answered]
Hi, I came across [this ](http://www.steves-internet-guide.com/creating-and-using-client-certificates-with-mqtt-and-mosquitto/)and found it very interesting. Does Tasmota support something like this or is there anything planned in this direction?

I'm imagining being able to subsequently exclude individual/separate devices from the MQTT broker and, above all, not having to use the same access data on every device.

**Comment by sfromis** [comment, marked as answer]
Not sure what you mean by "something like", but Tasmota does support [MQTT-TLS](https://tasmota.github.io/docs/TLS/).

**Reply by ChrisVonFrosta** [reply]
sorry for unclear question. i was looking for an option to do mqtt-authentication without username/password, rather use different certificates for different clients (tasmota devices). I am playing arround with mosqutitto on my public server.

Thank you for the Link [MQTT-TLS](https://tasmota.github.io/docs/TLS/). I am not sure if i understand that feature correctly. Even both options. 
But this is to authenticate without username and password right?
I need more time and translation to dive deeper. I was confused from the instructions and options.
"Auto-learn" sounds easy but it is not clear for me, what needs on the server side in comparison to other option.
Farthermore, the instruction is talking about the fingerprint tool and simultaniusly that this tool is deprecated.
But it is still needed for first fingerprint creation?
Fingerprint is an abstract or result depending from the servers private key?

By reading  [MQTT-TLS](https://tasmota.github.io/docs/TLS/) i came around to [Self singed Mosquitto](https://tasmota.github.io/docs/Self-signed-Mosquitto/). So the first is Tasmota side and the second is depending the server. It a lot of information and more activity necessary than i imagined before.

You can ignore my questions if you think, i did not spend enough time do dive in.

That's how I understood it so far:
I create the pair of keys. preferably on a third system.
copy keys pair to the server and config the mosquitto.
Compile Tasmota with TLS/Fingerprint options and create the initial fingerprint with the tool and copy it to tasmota or use auto-learn.
But on auto-learn i would missing the authentication against or through the server. I think i am still in the fog xD

**Reply by barbudor** [reply]
MQTT-TLS as nothing to do with authentication but encryption of the MQTT protocol just like HTTPS is encrypted HTTP.

What Steve's describe and what you seems to be looking for, is key based authentication, like in SSH.
AFAIK, Tasmota doesn't support that or at least I don't remember seeing any such discussion.
Unless may be the AWS IOT protocol use such kind of authentication ?
May be @s-hadinger would have some element to share

**Reply by ChrisVonFrosta** [reply]
Sure. Thank you for shedding some light on my confusion.

**Reply by s-hadinger** [reply]
Wise advice, get out of certificate authentication. Given that there is no secure storage in Tasmota, passwords are not less secure

**Reply by ChrisVonFrosta** [reply]
that was less from the sercure view, more from the option to have client-sepcific login data.

**Reply by barbudor** [reply]
What do you mean ?

When using user/password authentication, you can use individual credentials per device
Each of my device connects to the broker using a different user (actually identical to the topic) and each has an indivdualized password

I also use ACL to control which device can read or write which topic. With user = topic it is easy to create ACL rules using `%u` within the topic patterns. These allows any tasmota to read (subscribe) any topic but write (publish) only to some topics:
```
pattern read  #
pattern write stat/%u/#
pattern write cmnd/%u/#
pattern write tele/%u/#
pattern write tasmota/discovery/#
```

**Reply by ChrisVonFrosta** [reply]
Since I always used ioBroker as an MQTT broker, I was never aware that I could have multiple users/passwords. While managing certificates still seems easier to me than organizing multiple usernames/passwords, it appears it's not meant to be.

Thx for the hint with ACL. i found that [page](https://tasmota.github.io/docs/Securing-your-IoT-from-hacking/) by google. I will check.

**Reply by s-hadinger** [reply]
Certificates are waaayyy more complex than passwords!
